CenterForOpenScience/osf.io
# [IN-229][OSF] Add email hooks and templates for reviews notifications.

## Purpose

Currently, no emails are sent to moderators/admins of a moderated preprint service upon preprint submission. This PR adds the functionality to automatically send emails to notify moderators/admins depending on their notification settings.

## Changes

Some background about the digest email workflow is needed for better understanding:
There are three settings for an `OSFUser`'s notification: instant, daily and none. `Instant` means the user would receive notification in digest form every five minutes, while `daily` means the user would be notified everyday at noon (also in digest form) if there are new activities. 
Two cron tasks, `5-minute-emails` and `daily-emails` (see `website/settings/defaults.py`), are scheduled so that celery would periodically run `send_users_email('email_transactional')` and `send_users_email('email_digest')`, repesctively.
When the celery task runs, it would try to fetch messages previously stored to be sent as digest by calling `get_user_emails()`, which would return an interable containing information derived from existing `NotificationDigest` instances, which were created and saved by calling `store_emails()`. These instances will be deleted at the end of the task by calling `remove_notifications()`.

In short, the workflow looks like this:
`Action triggers a notification -> NotificationDigest instance is stored -> celery task periodically runs and retrieves information from existing NotificationDigest instances -> Digest email is compiled and sent -> NotificationDigest instances are deleted`

Therefore, the following changes are made, in the order of the workflow described above.
1. In `website/reviews/listeners.py`, a signal is added to invoke `store_emails()` when new preprint is submitted.
    - In `osf/utils/machines.py`, the signal is sent in `notify_submit()`
    - `store_emails()` has a new param `provider` so that it can store provider relationship with the newly added `provider` field of the `NotificationDigest` model.
2. In `website/notifications/tasks.py`, `send_users_email()` are modified so that it fetches all reviews-related email by calling `get_moderator_emails()` and send them.
    -  `get_moderator_emails()` is added to return moderator related emails.
    - Query string of `get_user_emails()` is modified so that it excludes moderator related emails for better decoupling.
3. New email templates are added. Existing templates are modified to show correct footer.

## QA Notes

Since this modifies `store_email()`, check and see if other digest notifications are still working.
Make sure that emails are sent to moderators at the right time (at five-minute intervals and everyday at noon).

## Side Effects

None

## Ticket

https://openscience.atlassian.net/browse/IN-229
